README.md

```diff
@@ -168,6 +168,9 @@ ktlint {
         exclude("**/generated/**")
         include("**/kotlin/**")
     }
+    customReporters {
```

Comment:
Possibly this syntax may be used for built-in reporters as well:
```gradle
reporters {
   repoter "plain"
   customReporter "csv", "csv", project(":csv-reporter")
}
```

Reply:
@JLLeitschuh any opinion on this?

Reply:
Can you explain them both better please so I know what I'm picking?

Reply:
Currently to set built-in reporter plugin user calls:
```kotlin
    reporters.set(setOf(ReporterType.PLAIN, ReporterType.CHECKSTYLE))
```

My idea is to move this to `reportersDsl { .. }` block, so setting reporters would be unified (see previous comment how it may look).

Reply:
Does the DSL also allow the developer to describe what the reporter's output type is?

Reply:
For built-in reporters developer can't define output type, same as it is already via `reporters.set(..)` call.

Reply:
Can you give an example of the DSL usage with only a single custom reporter added?

Reply:
Only one custom reporter added:
```gradle
reporters {
   customReporter "csv", "csv", project(":csv-reporter")
}
```

Only built-in reporter added:
```gradle
reporters {
   repoter "plain"
}
```

Reply:
In the documentation can we make it:

```kotlin
reporters {
   customReporter "csv-reporter", ".csv", project(":csv-reporter")
}
```

That way it's really clear what's going on for each parameter?

Reply:
Also, you probably want to accept a `Configuration` as the last argument instead of a project, right?

Reply:
First parameter of custom reporter is the id of the reporter, so it should stay "csv": https://github.com/Tapchicoma/ktlint-gradle/blob/ce38cea8af5654c60b924329f4d353d1aac7c042/samples/kotlin-reporter-creating/src/main/kotlin/org/jlleitschuh/gradle/ktlint/sample/CsvReporterProvider.kt#L8

Second parameter is generated reporter file extension, it should always contain '.' symbol, so, imho, better to omit it.

last parameter is "dependency notation" from Gradle [DependencyHandler](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/dsl/DependencyHandler.html), that is added by plugin to internal "ktlintReporter" configuration.

Reply:
I guess the point that I'm trying to make is that from the perspective of a lay-person, the API is somewhat confusing and unclear what each parameter is.

May I suggest an alternative API?

```kotlin
reporters {
    customReporters {
        "csv" {
            extension = "csv" // This could also be implied from the name
            from(project(":csv-reporter"))
        }
    }
}
```

In order to accomplish this API, you could make the `customReporters` a `NamedDomainObjectContainer`.

Simply making `customReporters` a `NamedDomainObjectContainer<CustomReporter>` where `CustomReporter` is something like:

```kotlin
data class CustomReporter(
    val name: String,
    var extension: String = name,
    var from: [Something]? = null
)
```

I know this is a lot to change, but I believe the API is clearer, more flexible, and is more "Gradle like".

Reply:
Hm, looks interesting. Let me play with it. I suppose built-in reporters would have proposed above syntax:
```kotlin
reporters {
    reporter "checkstyle"
}
```
?

Reply:
You could keep the built in ones as simple as you want.

Reply:
@JLLeitschuh please check this commit https://github.com/JLLeitschuh/ktlint-gradle/pull/258/commits/33d510c5323deaeef1d8a8afa83ea07ef9ad0dc8 if this new DSL looks good for you.